### PR TITLE
evaluator: move the evaluate of ast functions to buildin

### PR DESCRIPTION
**Pull request by zxylvlp**
Move the evaluate of ast functions extract, convert and substring to buildin.

### CODE REVIEW

**Comment by zxylvlp** [comment]
PTAL @coocood  @zimulala @ngaut

File: evaluator/builtin/string.go [hunk 1]
```diff
@@ -188,3 +191,85 @@ func builtinReplace(args []interface{}, _ context.Context) (interface{}, error)
...
+
+	t := args[1]
+	p, ok := t.(int64)
+	if !ok {
+		return nil, errors.Errorf("Substring invalid pos args, need int but get %T", t)
+	}
+	pos := int(p)
+
+	length := -1
+	if len(args) >= 3 {
```
Comment (coocood) [inline evaluator/builtin/string.go]: This make me wondering when will `len(args)` be greater than 3, but actually `len(args)` will never be greater than 3.

Comment (zxylvlp) [inline evaluator/builtin/string.go]: OK, I will change it to 'len(args)==3'.

File: evaluator/builtin/builtin.go [hunk 2]
```diff
@@ -87,6 +88,8 @@ var Funcs = map[string]Func{
 	"replace":   {builtinReplace, 3, 3, true, false},
 	"upper":     {builtinUpper, 1, 1, true, false},
 	"strcmp":    {builtinStrcmp, 2, 2, true, false},
+	"convert":   {builtinConvert, 2, 2, true, false},
+	"substring": {builtinSubstring, 2, 3, true, false},
```
Comment (shenli) [inline evaluator/builtin/builtin.go]: The 4th parameter is "IsStatic". For substring, it is not always static.

File: optimizer/typeinferer.go [hunk 3]
```diff
@@ -267,13 +267,13 @@ func (v *typeInferrer) handleFuncCallExpr(x *ast.FuncCallExpr) {
 		tp = types.NewFieldType(mysql.TypeDatetime)
 	case "microsecond", "second", "minute", "hour", "day", "week", "month", "year",
 		"dayofweek", "dayofmonth", "dayofyear", "weekday", "weekofyear", "yearweek",
-		"found_rows", "length":
+		"found_rows", "length", "extract":
 		tp = types.NewFieldType(mysql.TypeLonglong)
 	case "now", "sysdate":
 		tp = types.NewFieldType(mysql.TypeDatetime)
 		tp.Decimal = v.getFsp(x)
 	case "dayname", "version", "database", "user", "current_user",
```
Comment (shenli) [inline optimizer/typeinferer.go]: miss substring?

File: ast/flag.go [hunk 4]
```diff
@@ -170,14 +164,6 @@ func (f *flagSetter) funcCall(x *FuncCallExpr) {
 	x.SetFlag(flag)
 }
 
-func (f *flagSetter) funcSubstring(x *FuncSubstringExpr) {
```
Comment (shenli) [inline ast/flag.go]: Should we set flag for builtin functions?

File: evaluator/builtin/builtin.go (hunk 2 above)
Comment (coocood) [inline evaluator/builtin/builtin.go]: The `IsStatic` and `IsAggregate` fields are not used any more, can be deleted.

**Comment by zxylvlp** [comment]
@shenli @coocood PTAL

**Comment by coocood** [comment]
LGTM

**Comment by shenli** [comment]
LGTM

**Pull request merged**